aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -567,5 +568,258 @@ Tensor& index_add_cuda_(Tensor & self, int64_t dim, const Tensor & index, const
+
+      IndexType srcOffset =
+        cuda::detail::IndexToOffset<T, IndexType, SrcDim>::get(linearIndex, src);
+      srcOffset += srcIndex * src.strides[srcSelectDim];
+
+      dst.data[dstOffset] = src.data[srcOffset];
+    }
+  }
+}
+
+// We prefer this kernel to balance parallelism across index points,
+// if there are a large number of indices.
+// This kernel in fact works for all choices of problem size, but if
+// the number of indices chosen is small, then the
+// indexSelectSmallIndex kernel is a better choice to reduce memory
+// accesses.
+// TODO: remove
+template <typename T, typename IndexType, int DstDim, int SrcDim, int IdxDim,
+          bool IndexIsMajor>
+__global__ void indexSelectLargeIndex(cuda::detail::TensorInfo<T, IndexType> dst,
+                                      cuda::detail::TensorInfo<T, IndexType> src,
+                                      cuda::detail::TensorInfo<int64_t, IndexType> indices,
+                                      int dstSelectDim,
+                                      int srcSelectDim,
+                                      IndexType totalSize,
+                                      IndexType innerSize,
+                                      int64_t srcSelectDimSize) {
+  // We stride over the output including the indexed dimension
+  // (totalSize), and calculate the destination index point based on that
+  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < totalSize;
+       linearIndex += gridDim.x * blockDim.x) {
+    IndexType dstIndex, elementInSlice;
+    if (IndexIsMajor) {
+      dstIndex = linearIndex / innerSize;
+      elementInSlice = linearIndex % innerSize;
+    }
+    else {
+      elementInSlice = linearIndex / innerSize;
+      dstIndex = linearIndex % innerSize;
+    }
+
+    // Lua indices begin at 1
+    IndexType srcIndex =
+      indices.data[cuda::detail::IndexToOffset<int64_t, IndexType, IdxDim>::get(dstIndex, indices)];
+    CUDA_KERNEL_ASSERT(srcIndex < srcSelectDimSize);
+
+    IndexType dstOffset =
+      cuda::detail::IndexToOffset<T, IndexType, DstDim>::get(elementInSlice, dst);
+    dstOffset += dstIndex * dst.strides[dstSelectDim];
+
+    IndexType srcOffset =
+      cuda::detail::IndexToOffset<T, IndexType, SrcDim>::get(elementInSlice, src);
+    srcOffset += srcIndex * src.strides[srcSelectDim];
+
+    dst.data[dstOffset] = src.data[srcOffset];
+  }
+}
+template<typename scalar_t>
+void index_select_out_cuda_impl(Tensor& out, const Tensor& self, long dim,
+                                const Tensor& index) {
+  ptrdiff_t numIndices = index.numel();
+
+  int selfDims = self.dim() == 0 ? 1 : self.dim();
+
+  const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  TORCH_CHECK(index.dim() <= 1,
+             "Index is supposed to be an empty tensor or a vector");
+  TORCH_CHECK(dim < selfDims, "Indexing dim is out of bounds");
+  TORCH_CHECK(selfDims > 0, "Source tensor is empty");
```

Comment:
Seems like this should use numel, since it's checking for whether the tensor is empty and tensors with non-zero dims can be empty (e.g. a tensor of shape [3, 0, 1]).

Disallowing this operation on empty tensors seems OK for now but in a future PR we may want to allow them. That is, a tensor with dims [3, 0, 1] can be validly selected on its left most dimension with [0, 2] to become a tensor with shape [2, 0, 1].

Reply:
My mistake, the current `index_select` works fine on tensors with no elements:

```
t = torch.randn((3, 0, 3))
t.index_select(0, torch.tensor([1, 1, 1, 1, 1]))
: tensor([], size=(5, 0, 3))
```

Is this check still relevant?

Reply:
Yes, this selfDims check can be dropped, good catch.